### ISSUE TITLE
Containers intermittently start with host /etc/resolv.conf instead of the kubernetes created file

### ISSUE THREAD

**Issue by matt-duch** [kubernetes/kubernetes#52655]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
The issue was noticed when some containers stopped being able to resolve cluster addresses. After a docker daemon restart, ~10% of containers that came up had invalid /etc/resolv.conf files (did not contain kubedns settings, just the host /etc/resolv/conf). This was distributed evenly across the 4 types of images/pod configurations in use. 

**What you expected to happen**:
All containers should start with a resolv.conf that matches the environment.

**How to reproduce it (as minimally and precisely as possible)**:
sudo systemctl restart docker
These nodes had ~80 pods each

**Anything else we need to know?**:
This may or may not be at all relevant (it could just be a coincidence):
This issue was noticed on 2 servers whose docker daemon was restarted, and the result was 9 containers with this issue, on each node (of an original ~80). Stopping the 9 containers with issues (docker stop ...) resulted in 1 container with this issue, on each node. Stopping that 1 container resulted in no containers with this issue. 

Printed container id + resolv.conf with:
sudo find /var/lib/docker/containers/ -name resolv.conf -exec sh -c 'echo "$0" && cat "$0"'  {} \; 
Counted containers with issue with (ignoring 1 for the kube-proxy container):
sudo find /var/lib/docker/containers/ -name resolv.conf -exec cat {} \; | grep 10.0.0.2

```
//kube-proxy resolv.conf
nameserver 10.0.0.2
search ec2.internal

//expected resolv.conf for other pods
nameserver 10.128.0.10
search default.svc.cluster.local svc.cluster.local cluster.local ec2.internal
options ndots:5

// invalid resolv.conf
# This file is managed by man:systemd-resolved(8). Do not edit.
#
# This is a dynamic resolv.conf file for connecting local clients directly to
# all known DNS servers.
#
# Third party programs must not access this file directly, but only through the
# symlink at /etc/resolv.conf. To manage man:resolv.conf(5) in a different way,
# replace this symlink by a static file or a different symlink.
#
# See man:systemd-resolved.service(8) for details about the supported modes of
# operation for /etc/resolv.conf.

nameserver 10.0.0.2
search ec2.internal
```

**Environment**:
- Kubernetes version (use `kubectl version`): 1.7.2
- Cloud provider or hardware configuration**: AWS EC2
- OS (e.g. from /etc/os-release): Container Linux by CoreOS stable (1409.7.0)
- Kernel (e.g. `uname -a`):Linux 4.11.11-coreos
- Install tools: manual download of kubelet 1.7.2
- Others: Docker version 1.12.6, build a82d35e

**Comment by matt-duch** [comment]
@kubernetes/sig-node-bugs

**Comment by lichuqiang** [comment]
Pod started with `hostNetwork` would use the `resolv.conf` of hosts, you can check the `hostNetwork` item of the abnormal pods first

**Comment by matt-duch** [comment]
Agreed. I checked 3 or 4 manually, and confirmed that there were pods with abnormal resolv.confs that did not use hostNetwork. As an additional sanity check, this cluster only has 6 pods with hostNetwork, which is less than the 18 abnormal pods I saw after the daemon restart.

**Comment by gg7** [comment]
/remove-lifecycle stale
/remove-lifecycle rotten

We have also had this happen after a network issue which broke etcd and made all Kubernetes masters unavailable.

Environment:

* Kubernetes version: 1.6.13
* Cloud provider or hardware configuration: Baremetal servers
* OS: Ubuntu 16.04, 64 bit
* Kernel: `4.4.0-9[678]`
* Install tools: custom Ansible playbooks
* Others: Docker version 1.11.2

**Comment by jfpucheu** [comment]
Hello, 

We have the same issue on 1.6.13 and 1.7.9 sometime the pods get resolv.conf from the node instead of local K8S cluster DNS.

Same issue here : https://github.com/kubernetes-incubator/kubespray/issues/1951

Jeff

**Comment by robdewit** [comment]
@jfpucheu Did you somehow find the cause of this problem? I seem to have the same issues on bare metal in 1.6.13 - pods get node's /etc/resolv.conf without cluster dns and adjusted search path.

**Comment by gjcarneiro** [comment]
I don't know if this helps but we used to have this problem with some frequency in our (1.6.x) cluster, but once we upgraded all the servers to have SSD disks (at least for the docker/kubernetes filesystems) the problem disappeared.  

I reckon this is some sort of race condition, or else a timeout due to slow disk followed by a error recovery code path that is not very well tested.  But it's just speculation...  the code is hard to follow..

Do your servers have SSDs, @jfpucheu?

**Comment by robdewit** [comment]
I think I have found the cause for this (in our situation at least).
We have /var/lib/docker symlinked to another file system and Kubelet running in docker container.
When Kubelet starts the pod container, it checks the resolv.conf in the container directory. This directory is expanded, so /var/lib/docker become /mnt/volume/docker. That path was not mounted into the Kubelet container so then Kubelet can't fix the resolv.conf and the docker default remains.
We saw the error messages in Kubernetes v1.5.6, but it was just a harmless warning, somehow it retried. Starting with 1.6 Kubelet's behaviour changed and it fails to work constantly. 
Having the exact path available in the Kubelet container fixed it for us.

Nice hint (that is written off as a harmless warning all over the place) is:
```
docker_manager.go:2282] Failed to create pod infra container: RunContainerError; Skipping pod "PODNAME(PODHASH)": ResolvConfPath "/mnt/vol/docker/containers/CONTAINERHASH/resolv.conf" does not exist
E0523 08:55:46.470666   14438 pod_workers.go:182] Error syncing pod PODHASH ("PODNAME(PODHASH)"), skipping: failed to "StartContainer" for "POD" with RunContainerError: "ResolvConfPath \"/mnt/vol/docker/containers/CONTAINERHASH/resolv.conf\" does not exist"
```

**Comment by ReSearchITEng** [comment]
This was our case as well, with OpenShift Enterprise 3.9 !
removing the link and mount disk directly to /var/lib/docker solved the resolve.conf issue!


> E0623 13:35:46.048553  113497 remote_runtime.go:92] RunPodSandbox from runtime service failed: rpc error: code = Unknown desc = rewrite resolv.conf failed for pod "registry-console-1-deploy": ResolvConfPath "/**mnt**/docker/containers/fcbf7c62895e060885f167f90858e62687c5d3f307ebb519c5e0c0e7978d3f1d/resolv.conf" does not exist

**Comment by Eloahman** [comment]
@robdewit 
You have got the point.

**Comment by gigi-at-zymergen** [comment]
Ran into this as well, and was able to switch from an `ln` to a `mount --bind` based on @robdewit 's summary of the issue and @memwey's pithy guidance . Was wondering if this guidance can be found anywhere in the documentation? 
Many people put the docker dir on larger disk mounts than the boot drive and probably assume a symlink is sufficient. Would be nice to understand this aspect of kubelet while setting up the cluster nodes.
